### ISSUE TITLE
Support other package sets

### ISSUE THREAD

**Issue by infinisil** [NixOS/nixpkgs-vet#31]
We want to use the `pkgs/by-name` convention for other package sets, e.g. https://github.com/NixOS/nixpkgs/pull/299347.

Here's a rough idea how this could look like:

## Package set config

Define some Nix config file to determine which package sets exist:

```nix
# pkgs/top-level/by-name-sets-config.nix
{
  packageSets = {
    toplevel = {
      attrPath = [ ];
      path = ./pkgs/by-name;
      shardLength = 2;
    };
    mySet = {
      # Any attrpath that will contain the attributes
      attrPath = [ "mySet" ];
      path = ./pkgs/mySet/by-name;
      # No sharding
      shardLength = null;
    };
  };
}
```

## Read the packages in Nix

Create a Nix file to automatically call all of them like the current [`by-name-overlay.nix`](https://github.com/NixOS/nixpkgs/blob/4c731c3c34b5e79c68c6210c827097cb284b80ab/pkgs/top-level/by-name-overlay.nix):

```nix
# pkgs/top-level/by-name-sets.nix
{
  toplevel = callPackage: {
    hello = <...>;
  };
  mySet = callPackage: {
    hi = <...>;
  };
}
```

For each package set declared, define it by importing `by-name-sets.nix`, like

```nix
mySet = lib.makeScope pkgs.newScope (self:
  (import ./pkgs/top-level/by-name-sets.nix).mySet self.callPackage
) 
```

## Check the packages with the tool

Adjust this tool to read a `by-name-sets-config.nix` file for determining what it needs to check and how.

**Comment by infinisil** [comment]
Had a chat today with @Dmills27, who's interested in contributing. I took some notes to break down this task:

1. Make [`by-name-overlay.nix`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/by-name-overlay.nix) be synced with the tool
   - Move `by-name-overlay.nix` into nixpkgs-check-by-name
   - Make sure it can accept a config argument in the future
   - Distribute this file in the releases of nixpkgs-check-by-name
     - Not necessary, Github always distributes the source tarball
   - In Nixpkgs, the [`update-pinned-tool.sh`](https://github.com/NixOS/nixpkgs/blob/051f920625ab5aabe37c920346e3e69d7d34400e/pkgs/test/check-by-name/update-pinned-tool.sh) should be updated to also fetch the Nix file from the release.
2. Add a config file with a configurable prefix length
   - Make the tool read a `config.json` (given as a CLI arg) and change behavior depending on it
   - Make `by-name-overlay.nix` read a `config.json` and change behavior depending on
   - See [here](https://github.com/NixOS/nix/blob/e1b3716d50101c3512fa55ad703169af76a11f71/doc/manual/src/contributing/json-guideline.md) for how to design the JSON
     ```json
     {
       "shard_length": 2
     }
     ```
   - In the future perhaps a module along with it, not necessary though
     ```nix
     {
       options.shard_length = lib.mkOption {
         type = nullOr int;
         # default = 2;
       }
     }
     ```

**Comment by fgaz** [comment]
+1! Here is another by-name package set: https://github.com/NixOS/nixpkgs/pull/344716

**Comment by Sigmanificient** [comment]
- Any news about this?

**Comment by Sigmanificient** [comment]
that would be awesome 👀 
*my preivous was hidden as abuse?* 🤔

**Comment by wolfgangwalther** [comment]
> _my preivous was hidden as abuse?_ 🤔

Yes, that was me. Any "+1"  or "any update on this" or anything along those lines, is not helping the topic progress and is just (trying to) creating implicit pressure on the respective maintainers. Since we're all working voluntarily on these open source projects, I consider these kinds of comments abuse.

**Comment by Sigmanificient** [comment]
@wolfgangwalther thanks for the clarification, and sorry, it was not my intention. I was wondering whether there was active work on it, not to pressure anyone, only gathering information. And potentially help too, if that's possible!
